### PR TITLE
Kubernetes discovery adapter rework

### PR DESCRIPTION
**Pull request by zcahana**
This PR reworks the existing Kubernetes adapter code, with the following key changes:

- The official [github.com/kubernetes/client-go](https://github.com/kubernetes/client-go) client package is now used, instead of a simple HTTP-based client. The official client includes support for watches, self-populating caches, API objects, etc.
- The adapter logic is now based on watches rather than polling (i.e., almost immediate response to changes)
- Instance tags are based on pod labels, instead of pod names conventions. Each label is mapped into a corresponding "key=value" tag. Dynamic label updates are watched and reflected as well. Additionally, pod annotations are mapped to instance metadata.
- The adapter can read URL/token configuration from the pod [service account](http://kubernetes.io/docs/user-guide/service-accounts/) (when running within a Kubernetes pod), instead of passing explicit configuration.
- The adapter can be configured to target a predefined namespace (defaults to "default" namespace).

Note: the `github.com/kubernetes/client-go` package(s), combined with its own dependencies, is very large (50+ MB). This also affect the compile time for the registry/sidecar (which use the adapter), as well as their produced binary sizes (~47MB). Pending PR #418 takes care of significantly shrinking the binaries.

### CODE REVIEW

**Comment by zcahana** [comment]
Note to reviewers: the `registry/adapters/kubernetes/kubernetes.go` file is essentially rewritten from scratch, so I would suggest to review it as a whole, instead of in the "diff" view.

File: pkg/datastructures/stringset.go
Comment (rshriram) [inline pkg/datastructures/stringset.go]: These functions seem arbitrary in this file. Why can't we use math.min and math.max ?

File: registry/adapters/kubernetes/kubernetes.go
Comment (rshriram) [inline registry/adapters/kubernetes/kubernetes.go]: what is the difference between these two? The comments seem to be the same for both.

File: registry/adapters/kubernetes/kubernetes.go
Comment (rshriram) [inline registry/adapters/kubernetes/kubernetes.go]: What is the difference between ServiceInstance and servicePods ?

File: registry/adapters/kubernetes/kubernetes.go
Comment (rshriram) [inline registry/adapters/kubernetes/kubernetes.go]: Once again, what is the difference between the two functions, ListInstances and ListServiceInstances ?

File: registry/adapters/kubernetes/kubernetes.go
Comment (rshriram) [inline registry/adapters/kubernetes/kubernetes.go]: nit. copy paste error

**Review by rshriram — CHANGES_REQUESTED** [review]
Comments and questions inline. We also need to filter out what we cache locally, since we are looking at a lot of duplicated information across pods on the same host.

Updating the examples would also be helpful.
Simultaneously, it would be very very helpful if we could make the sidecar use this library to fetch K8S instances instead of querying the registry service.

File: pkg/datastructures/stringset.go
Comment (zcahana) [reply pkg/datastructures/stringset.go]: math.Min/Max work with floats, so are inconvenient and inefficient for use with integers.
We can, however, move these to our own `amalgam8/pkg/math` package, to be reused across the project.

File: registry/adapters/kubernetes/kubernetes.go
Comment (zcahana) [reply registry/adapters/kubernetes/kubernetes.go]: first one is used for the endpoints cache, and the second one for the pods cache. I'll fix the comments.

File: registry/adapters/kubernetes/kubernetes.go
Comment (zcahana) [reply registry/adapters/kubernetes/kubernetes.go]: `api.ServiceInstance` is a8's representation of an instance of a service. It is built and stored in the `services` map, so that it won't have to be recalculated with every `ListInstances()` call.
`servicePods` maps service name to a set of pod UIDs that implement it. This is maintained, together with the opposite mapping (pod UID --> services), so that we can supporting dynamic labels update on pods.

File: registry/adapters/kubernetes/kubernetes.go
Comment (zcahana) [reply registry/adapters/kubernetes/kubernetes.go]: One list all instances, and the other one list instances for a particular service.
This is part of the [`api.ServiceDiscovery`](https://github.com/amalgam8/amalgam8/blob/master/registry/api/types.go#L24) interface, so required to be here in order to implement it.

**Comment by zcahana** [comment]
@rshriram I'm already working on migrating the (k8s) examples to use the adapter. This will be done on a separate PR. However, merging it will have to wait until the adapter rework is released, and the updated examples images are in Dockerhub.

Also, the sidecar can already use the adapter to talk to k8s directly. This was enabled in PR #397 a while ago.

File: registry/adapters/kubernetes/kubernetes.go
Comment (rshriram) [reply registry/adapters/kubernetes/kubernetes.go]: Excellent. Can you just stick this into the comments please?

File: pkg/datastructures/stringset.go
Comment (rshriram) [reply pkg/datastructures/stringset.go]: Is it more inconvenient than maintaining our own version of the math package for integers? The type casting inefficiency is not so high. This is micro optimization considering all the string operations we are doing. But I leave it to you.

File: registry/adapters/kubernetes/kubernetes.go
Comment (rshriram) [reply registry/adapters/kubernetes/kubernetes.go]: Once again, stick this into comments please.

**Comment by rshriram** [comment]
Yes I am aware of PR #397. However I was hoping a zero config example with this PR (i.e. read from service accounts). 

Let's get this PR fixed and merged. As you said, the examples would depend on this PR. 

On a side note: do we need explicit configuration for k8s URL at all ? According to google docs, the client was supposed to be able to "automatically" connect to the API server (could find out exactly where the magic of deciphering the k8s API server IP address  was happening). 

If no explicit configuration is needed, can we get rid of related code? 

Secondly, please update the docs as well. Else we would end up doing a major rework of docs for the full k8s feature set.

File: registry/adapters/kubernetes/kubernetes.go
Comment (zcahana) [reply registry/adapters/kubernetes/kubernetes.go]: That's... already there :)

Comment (rshriram) [reply registry/adapters/kubernetes/kubernetes.go]: Oh. nm my bad.

**Comment by zcahana** [comment]
Zero* configuration is already supported in this PR: the adapter can automatically fetch the K8s URL/Token from the service account. It also supports explicit URL/Token configuration, so that it can run out of a K8s pod.

\* Zero is actually not zero, because you still need to configure *what* type of registry you're working with - be it A8 registry, Eureka or K8s. But with K8s, that's all you need to configure.

**Comment by zcahana** [comment]
More updates underway, do not merge.

File: pkg/datastructures/stringset.go
Comment (frankbu) [reply pkg/datastructures/stringset.go]: I think Zvi's approach of just making his own simple min/max functions is typical: http://mrekucci.blogspot.ca/2015/07/dont-abuse-mathmax-mathmin.html

**Comment by zcahana** [comment]
Add 3 commits addressing code review comments from @zachidan and @rshriram:
- Move min/max functions to their own package
- Switch to use a lower-level RESTClient
- Bugfix in bookkeeping of the podServices data structure
- Index pods by their names rather than by UIDs
- Fetch cached pods/endpoints by key rather than by linear search
- Comments fix
- Use a regular Mutex instead of RWMutex (single reader expected)
- Use finer-grain locking
- Use a single-worker workqueue to process watch events
- Avoid re-processing endpoints/pod updates when resource version hasn't changed (on full resync)

Also rebased the branch on master (to get the fix for issue #401), which seems to mess a bit with t the commits order.

**Comment by zcahana** [comment]
Two additional notes:
1. I've looked into filtering out pod data from our cache, so that we keep only what we need (names, labels and annotations). Apparently, when using the build-in k8s client cache, there's no easy way to do so, if at all. The alternative is probably to stop use that cache, and maintain our own. So, to keep things simple, let's live with the current approach for now.
2. The integration tests on Travis are failing because the registry, when linked with the k8s client packages,  doesn't produce a statically linked binary, which fails to start on a "scratch"-based container. PR #418 includes a fix to our compile/links flags, so that we get a proper static linked binary. Once this is merged, we can rebase this branch and get a green build.

**Review by rshriram — APPROVED** [review]
Looks good for the initial pass.

**Comment by rshriram** [comment]
@zcahana the PR looks good for first pass. Seems the branch is out of sync (as per GitHub UI). Do whatever sync you need to and merge it.

**Comment by rshriram** [comment]
fixes #172, fixes #173

**Pull request merged**
